### PR TITLE
Gossip query extensions (EXPERIMENTAL_FEATURES)

### PR DESCRIPTION
**Pull request by rustyrussell**
~**DO NOT MERGE UNTIL RFC FINALIZED** (it's EXPERIMENTAL_FEATURES, so we could, but might as well wait until tomorrow's meeting).~ This was the agreed final form, w00t!

This brings is close to in line with https://github.com/lightningnetwork/lightning-rfc/pull/557

At least, we reproduce the test vectors.  We still need behavoural tests (ie. protocol tests), but the code itself is fairly straightforward.

### CODE REVIEW

File: wire/extracted_peer_experimental_csv [hunk 1]
```diff
@@ -23,7 +23,7 @@
  msgtype,reply_channel_range,264,gossip_queries
  msgdata,reply_channel_range,chain_hash,chain_hash,
  msgdata,reply_channel_range,first_blocknum,u32,
-@@ -168,6 +189,20 @@
+@@ -168,6 +189,19 @@
```
Comment (cdecker) [inline wire/extracted_peer_experimental_csv]: This doesn't look like it belongs here. Is this a diff-fragment?

File: gossipd/test/run-extended-info.c [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+void memleak_remove_intmap_(struct htable *memtable UNNEEDED, const struct intmap *m UNNEEDED)
+{ fprintf(stderr, "memleak_remove_intmap_ called!\n"); abort(); }
+#endif
+
+/* Generated stub for status_fmt */
+void status_fmt(enum log_level level UNNEEDED, const char *fmt UNNEEDED, ...)
+{
+}
+
+static const char *test_vectors[] = {
```
Comment (cdecker) [inline gossipd/test/run-extended-info.c]: Loading the test vectors from file would have allowed us to update them more easily in the future, but for now that's ok.

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - JSON API: `listfunds` now returns also `funding_output` for `channels`
 - plugins: plugins can now suggest `lightning-cli` default to -H for responses.
 - Plugin: new notification `forward_event` offered/settled/failed/local_failed.
+- Protocol: --enable-experimental-features support gossip query extensions proposal aka https://github.com/lightningnetwork/lightning-rfc/pull/557
```
Comment (cdecker) [inline CHANGELOG.md]:
```suggestion
- Protocol: `--enable-experimental-features` support gossip query extensions proposal aka https://github.com/lightningnetwork/lightning-rfc/pull/557
```

File: gossipd/test/run-extended-info.c [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+	u8 *msg;
+	size_t i;
+
+	get_chainhash(test_vector, obj, &chain_hash);
+	assert(json_to_number(test_vector, json_get_member(test_vector, obj, "firstBlockNum"), &firstBlockNum));
+	assert(json_to_number(test_vector, json_get_member(test_vector, obj, "numberOfBlocks"), &numberOfBlocks));
+	opt = json_get_member(test_vector, obj, "extensions");
+	json_for_each_arr(i, t, opt) {
+		assert(json_tok_streq(test_vector, t,
+				      "WANT_TIMESTAMPS | WANT_CHECKSUMS"));
```
Comment (niftynei) [inline gossipd/test/run-extended-info.c]: would it be worthwhile to also test these separately?

File: gossipd/test/run-extended-info.c (hunk 2 above)
Comment (rustyrussell) [reply gossipd/test/run-extended-info.c]: There'll be a rework of test vectors in the spec sometime, and we can clean up how we treat them then.

File: gossipd/gossipd.c [hunk 5]
```diff
@@ -945,23 +1092,67 @@ static bool queue_channel_ranges(struct peer *peer,
...
+			goto wont_fit;
+		/* 1 byte for encoding_type, too */
+		extension_bytes += 1 + tlv_len(tstamps->encoded_timestamps);
 	}
 
 	/* If we can encode that, fine: send it */
-	if (encoding_end_prepend_type(&encoded_scids, max_encoded_bytes)) {
+	if (extension_bytes <= max_encoded_bytes
+	    && encoding_end_prepend_type(&encoded_scids,
+					 max_encoded_bytes - extension_bytes)) {
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply gossipd/gossipd.c]: I don't understand the question?

File: gossipd/test/run-extended-info.c (hunk 4 above)
Comment (rustyrussell) [reply gossipd/test/run-extended-info.c]: Yes, but the test vectors are grossly insufficient.  We need behavoural tests

**Comment by rustyrussell** [comment]
Applied @cdecker fix and rebased for CHANGELOG.md conflict.

Ack https://github.com/ElementsProject/lightning/pull/2900/commits/c5ac1fe1d43d91dea3d6cfd5c633f81b4b9ebc58

**Pull request merged**